apps/root/src/hooks/useYieldOptions.ts

```diff
@@ -1,52 +1,22 @@
-import React from 'react';
-import isUndefined from 'lodash/isUndefined';
-import isEqual from 'lodash/isEqual';
-import usePrevious from '@hooks/usePrevious';
 import { YieldOptions } from '@types';
-import useSelectedNetwork from './useSelectedNetwork';
-import useYieldService from './useYieldService';
-import useAccount from './useAccount';
+import PairService, { PairServiceData } from '@services/pairService';
+import usePairService from './usePairService';
+import useServiceEvents from './useServiceEvents';
 
-function useYieldOptions(chainId?: number, useBlacklist = false): [YieldOptions | undefined, boolean, string?] {
-  const [isLoading, setIsLoading] = React.useState(false);
-  const yieldService = useYieldService();
-  const [result, setResult] = React.useState<YieldOptions | undefined>(undefined);
-  const [error, setError] = React.useState<string | undefined>(undefined);
-  const currentNetwork = useSelectedNetwork();
-  const chainIdToUse = chainId || currentNetwork.chainId;
-  const prevChainId = usePrevious(chainIdToUse);
-  const prevUseBlacklist = usePrevious(useBlacklist);
-  const prevResult = usePrevious(result, false);
-  const account = useAccount();
+function useYieldOptions(chainId?: number): [YieldOptions | undefined, boolean] {
+  const pairService = usePairService();
 
-  React.useEffect(() => {
-    async function callPromise() {
-      try {
-        const options = await yieldService.getYieldOptions(chainIdToUse, useBlacklist);
-        setResult(options);
-        setError(undefined);
-      } catch (e) {
-        setError(e as string);
-      }
+  const yieldOptions = useServiceEvents<PairServiceData, PairService, 'getYieldOptions'>(
+    pairService,
+    'getYieldOptions'
+  );
 
-      setIsLoading(false);
-    }
+  const hasFetchedAvailablePairs = useServiceEvents<PairServiceData, PairService, 'getHasFetchedAvailablePairs'>(
+    pairService,
+    'getHasFetchedAvailablePairs'
+  );
 
-    if (
-      (!isLoading && isUndefined(result) && !error) ||
-      !isEqual(prevChainId, chainIdToUse) ||
-      !isEqual(prevUseBlacklist, useBlacklist)
-    ) {
-      setIsLoading(true);
-      setResult(undefined);
-      setError(undefined);
-
-      // eslint-disable-next-line @typescript-eslint/no-floating-promises
-      callPromise();
-    }
-  }, [isLoading, result, error, account, currentNetwork, chainIdToUse, prevChainId, useBlacklist, prevUseBlacklist]);
-
-  return [result || prevResult, isLoading, error];
+  return [(chainId ? yieldOptions[chainId] : []) || [], !hasFetchedAvailablePairs];
```

Comment:
return type can be just YieldOptions instead of YieldOptions | undefined at the definition of the hook

Reply:
Not bad yeah